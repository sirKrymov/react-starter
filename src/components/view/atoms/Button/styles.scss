@import '../../../../styles/configs';

.button {
  border-radius: $border-radius;
  line-height: 1;
  transition: opacity $anim-time-fast, background-color $anim-time-fast,
    border-color $anim-time-fast, color $anim-time-fast;
  min-height: inherit;
  padding: 0 16px;
  width: 100%;

  &:active {
    opacity: 0.75;
  }

  &__content {
    align-items: center;
    display: inline-flex;
  }

  &--loading {
    pointer-events: none;
    cursor: default;

    &:hover {
      opacity: 1;
    }
  }

  &--color-secondary {
    background-color: var(--secondary);
    color: var(--secondaryContrastText);

    &.button--variant-outlined {
      background-color: transparent;
      border: 2px solid var(--secondary);
      color: var(--secondary);

      &:hover {
        background-color: var(--secondary);
        color: var(--secondaryContrastText);
      }
    }

    &.button--variant-contained {
      background-color: var(--secondary);
      box-shadow: $box-shadow-light;
      color: var(--secondaryContrastText);

      &:hover {
        background-color: var(--secondaryDark);
      }
    }
  }

  &--disabled {
    pointer-events: none;
    opacity: 0.6;
    cursor: not-allowed;
  }

  &--padding-xs {
    padding: 0 16px;
  }

  &--padding-sm {
    padding: 0 24px;
  }

  &--padding-md {
    padding: 0 32px;
  }

  &--padding-lg {
    padding: 0 40px;
  }

  &--font-weight-bold {
    font-weight: 700;
  }

  &--font-weight-default {
    font-weight: initial;
  }
}
