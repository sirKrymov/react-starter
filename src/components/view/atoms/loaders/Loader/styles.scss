.loader {
  border-radius: 50%;
  border-style: solid;
  animation: loaderAnim 0.8s infinite, show 0.4s;
  margin: auto;

  &--color-primary {
    border-color: var(--primary) transparent var(--primary) transparent;
  }

  &--color-secondary {
    border-color: var(--secondary) transparent var(--secondary) transparent;
  }

  &--color-on-secondary {
    border-color: var(--secondaryContrastText) transparent
      var(--secondaryContrastText) transparent;
  }

  &--size-sm {
    border-width: 2px;
    height: 26px;
    width: 26px;
  }

  &--size-md {
    border-width: 3px;
    height: 34px;
    width: 34px;
  }

  &--size-lg {
    border-width: 4px;
    height: 54px;
    width: 54px;
  }
}
