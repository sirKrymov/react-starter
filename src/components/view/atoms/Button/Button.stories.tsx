import React from 'react';

import { jsxDecorator } from 'storybook-addon-jsx';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import { Button } from '../Button';

import '../Button/styles.scss';

storiesOf('Button', module)
  .addDecorator(jsxDecorator)
  .add('Simple text', () => {
    return (
      <Button onClick={action('Clicked')} id="button">
        Hello Button
      </Button>
    );
  });
