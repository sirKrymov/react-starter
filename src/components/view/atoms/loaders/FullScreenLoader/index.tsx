import React, { ReactElement } from 'react';

import cn from 'classnames';

import './styles.scss';

interface IProps {
  className?: string;
  opacity?: boolean;
}

export function FullScreenLoader({ className, opacity }: IProps): ReactElement {
  const fullScreenLoaderClass = cn(
    {
      'full-screen-loader': true,
      'full-screen-loader--opacity': opacity
    },
    className
  );

  return (
    <div className={fullScreenLoaderClass}>
      <div className="container">
        <span className="layer" />
        <span className="layer" />
        <span className="layer" />
      </div>
    </div>
  );
}
