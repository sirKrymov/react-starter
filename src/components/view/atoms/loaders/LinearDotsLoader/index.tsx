import React, { ReactElement } from 'react';

import cn from 'classnames';

import './styles.scss';

interface IProps {
  className?: string;
  stretch?: boolean;
  color?: 'primary' | 'secondary';
  size?: 'sm' | 'md' | 'lg';
}

export function LinearDotsLoader({
  className,
  stretch,
  color,
  size
}: IProps): ReactElement {
  const linearDotsLoaderClass = cn(
    {
      'linear-dots-loader': true,
      [`linear-dots-loader--${color}`]: color,
      [`linear-dots-loader--${size}`]: size,
      'linear-dots-loader--stretch': stretch
    },
    className
  );

  return (
    <p className={linearDotsLoaderClass}>
      <span />
      <span />
      <span />
    </p>
  );
}
