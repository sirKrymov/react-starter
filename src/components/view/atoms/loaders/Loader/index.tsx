// @SETUP: Please don't forget to remove this component or current comment if component is used in project
import React, { ReactElement } from 'react';

import cn from 'classnames';

import './styles.scss';

export interface IProps {
  className?: string;
  color?: 'primary' | 'secondary' | 'on-secondary';
  style?: object;
  size?: 'sm' | 'md' | 'lg';
}

export function Loader({
  className,
  color = 'secondary',
  style,
  size = 'md'
}: IProps): ReactElement {
  const loaderClass = cn(
    {
      loader: true,
      [`loader--color-${color}`]: color,
      [`loader--size-${size}`]: size
    },
    className
  );

  return <div className={loaderClass} style={style} />;
}
