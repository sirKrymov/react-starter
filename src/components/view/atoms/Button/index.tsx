// @SETUP: Please don't forget to remove this component or current comment if component is used in project
import React, { CSSProperties, ReactElement, ReactNode, memo } from 'react';

import cn from 'classnames';

import { Typography } from 'components/view/atoms/Typography';
import { Loader } from 'components/view/atoms/loaders/Loader';
import { UiSize } from 'components/view/atoms/UiSize';

import './styles.scss';

interface IProps {
  fontWeight?: 'default' | 'bold';
  uppercase?: boolean;
  onClick?(): void | undefined;
  className?: string;
  disabled?: boolean;
  children: ReactNode;
  variant?: 'contained' | 'outlined';
  loading?: boolean;
  height?: 'xs' | 'sm' | 'md' | 'lg' | 'full';
  border?: boolean;
  width?: 'xs' | 'sm' | 'md' | 'lg' | 'xl' | 'full' | 'content';
  color?: 'secondary';
  style?: CSSProperties;
  type?: 'button' | 'submit';
  id: string;
}

function ButtonBase({
  fontWeight = 'bold',
  uppercase,
  className,
  disabled,
  children,
  onClick,
  variant = 'contained',
  loading,
  border,
  height,
  width,
  style,
  color = 'secondary',
  type = 'button',
  id
}: IProps): ReactElement {
  const buttonClass = cn(
    {
      button: true,
      [`button--font-weight-${fontWeight}`]: fontWeight,
      [`button--variant-${variant}`]: variant,
      [`button--padding-${height}`]: height && width === 'content',
      [`button--color-${color}`]: color,
      [`button--disabled`]: disabled,
      [`button--border`]: border,
      'button--loading': loading
    },
    className
  );

  return (
    <UiSize width={width} height={height}>
      <button
        className={buttonClass}
        disabled={disabled}
        onClick={loading || disabled ? undefined : onClick}
        style={style}
        type={type}
        id={id}
      >
        {loading ? (
          <Loader color="on-secondary" />
        ) : (
          <div className="button__content">
            <Typography uppercase={uppercase} variant="inherit" color="inherit">
              {children}
            </Typography>
          </div>
        )}
      </button>
    </UiSize>
  );
}

export const Button = memo(ButtonBase);
