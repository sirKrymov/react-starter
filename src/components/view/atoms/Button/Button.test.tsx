import React from 'react';

import { shallow, ShallowWrapper } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import sinon from 'sinon';

import { Loader } from '../loaders/Loader';
import { Button } from '../Button';

describe('<Button> atom component', () => {
  const handleClick = sinon.spy();

  let shallowComponent: ShallowWrapper;

  const createComponent = (props?: Record<string, any>) => {
    shallowComponent = shallow(
      <Button onClick={handleClick} id="button" {...props}>
        Test
      </Button>
    );
  };

  describe('WHEN the component loads', () => {
    beforeEach(() => {
      createComponent();
    });

    it('THEN it should render without crashing', () => {
      expect(shallowComponent.html()).not.toHaveLength(0);
    });

    it('THEN it should match initial snapshot', () => {
      expect(shallowToJson(shallowComponent)).toMatchSnapshot('Initial');
    });

    it('AND it should handle initial button logic', () => {
      shallowComponent.find('button').simulate('click');
      expect(handleClick).toHaveProperty('callCount', 1);
    });
  });

  describe('WHEN the component have loading proccess', () => {
    beforeEach(() => {
      createComponent({ loading: true });
    });

    it('THEN it should match snapshot with loading', () => {
      expect(shallowToJson(shallowComponent)).toMatchSnapshot('With loading');
    });

    it('AND it should show loading component', () => {
      expect(shallowComponent.find('button.button--loading')).toHaveLength(1);
      expect(shallowComponent.find(Loader)).toBeDefined();
    });
  });

  describe('WHEN the component have disabled status', () => {
    it('AND it should disable button if it contains disabled props', () => {
      createComponent({ disabled: true });

      expect(shallowComponent.find('button.button--disabled')).toHaveLength(1);
      shallowComponent.find('button').simulate('click');
      handleClick.calledOnceWith(undefined);
    });
  });
});
