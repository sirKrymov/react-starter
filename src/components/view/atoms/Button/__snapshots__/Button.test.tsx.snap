// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button> atom component WHEN the component have loading proccess THEN it should match snapshot with loading: With loading 1`] = `
<UiSize>
  <button
    className="button button--font-weight-bold button--variant-contained button--color-secondary button--loading"
    id="button"
    type="button"
  >
    <Loader
      color="on-secondary"
    />
  </button>
</UiSize>
`;

exports[`<Button> atom component WHEN the component loads THEN it should match initial snapshot: Initial 1`] = `
<UiSize>
  <button
    className="button button--font-weight-bold button--variant-contained button--color-secondary"
    id="button"
    onClick={[Function]}
    type="button"
  >
    <div
      className="button__content"
    >
      <Typography
        color="inherit"
        variant="inherit"
      >
        Test
      </Typography>
    </div>
  </button>
</UiSize>
`;
